src/proxy/mod.rs

```diff
@@ -0,0 +1,200 @@
+        }
+    }
+}
+
+/// Asynchronously generates and organizes payload data from an HTTP request based on the specified proxy type.
+/// This function ensures that requests are properly formatted to the correct service,
+/// returning a tuple with the request and the structured payload.
+pub(crate) async fn generate_payload_from_req(
+    req: Request<Body>,
+    proxy_type: &ProxyType,
+) -> GenericResult<(Request<Body>, PayloadData)> {
+    match proxy_type {
+        ProxyType::Quicknode => {
+            let (req, payload, signed_message) =
+                parse_body_payload::<QuicknodePayload>(req).await?;
+            let payload_data = PayloadData::Quicknode {
+                payload,
+                signed_message,
+            };
+            Ok((req, payload_data))
+        }
+        ProxyType::Moralis => {
+            let (req, signed_message) = parse_header_payload(req).await?;
+            Ok((req, PayloadData::Moralis(signed_message)))
+        }
+    }
+}
+
+pub(crate) async fn proxy(
+    cfg: &AppConfig,
+    req: Request<Body>,
+    remote_addr: &SocketAddr,
+    payload: PayloadData,
+    x_forwarded_for: HeaderValue,
+    proxy_route: &ProxyRoute,
+) -> GenericResult<Response<Body>> {
+    match payload {
+        PayloadData::Quicknode {
+            payload,
+            signed_message,
+        } => {
+            proxy_quicknode(
+                cfg,
+                req,
+                remote_addr,
+                payload,
+                signed_message,
+                x_forwarded_for,
+                proxy_route,
+            )
+            .await
+        }
+        PayloadData::Moralis(signed_message) => {
+            proxy_moralis(
+                cfg,
+                req,
+                remote_addr,
+                signed_message,
+                x_forwarded_for,
+                proxy_route,
+            )
+            .await
+        }
+    }
+}
+
+pub(crate) async fn validation_middleware(
+    cfg: &AppConfig,
+    payload: &PayloadData,
+    proxy_route: &ProxyRoute,
+    req_uri: &Uri,
+    remote_addr: &SocketAddr,
+) -> Result<(), StatusCode> {
+    match payload {
+        PayloadData::Quicknode { signed_message, .. } => {
+            validation_middleware_quicknode(cfg, signed_message, proxy_route, req_uri, remote_addr)
+                .await
+        }
+        PayloadData::Moralis(signed_message) => {
+            validation_middleware_moralis(cfg, signed_message, proxy_route, req_uri, remote_addr)
+                .await
+        }
+    }
+}
+
+/// Parses the request body and the `X-Auth-Payload` header into a payload and signed message.
+///
+/// This function extracts the `X-Auth-Payload` header from the request, parses it into a `SignedMessage`,
+/// and then reads and deserializes the request body into a specified type `T`.
+/// If the body is empty or the header is missing, an error is returned.
+async fn parse_body_payload<T>(
```

Comment:
Header is not part of the body, so function name and its documentation doesn't seem to be connected.

Reply:
changed namings for both parse functions https://github.com/KomodoPlatform/komodo-defi-proxy/pull/22/commits/e81cfb91ad505da4b36d23e55220d26850d904b8